File: AoC2020/solutions/09/solution.py
#Advent of Code 2020 day 9
from util import *
DAY = 9

def get_data():
    return [int(x) for x in input_lines(DAY)]

data = get_data()

first = None
data_range = 25



def solution():
    for i in range(data_range, len(data)):
        s = data[i-data_range:i]
        res = []
        for x in s:
            for y in s:
                if x != y:
                    res.append(x + y)

        if data[i] not in res:
            first = data[i]
            break

    second = None
    for i in range(len(data) - 1):
        sum = 0
        for j in range(i, len(data) - 1):
            sum += data[j]
            if sum > first:
                break
            if sum == first:
                d = sorted([data[x] for x in range(i, j+1)])
                second = d[0] + d[-1]
                break
        if second:
            break

    print(f"First:  {first}")
    print(f"Second: {second}")

solution()

File: AoC2020/solutions/09/solution.cpp
//Advent of Code 2020 day 9
#ifndef AOC_09_H
#define AOC_09_H
#include "util.hpp"
#include <algorithm>

std::string sourceDirectory = "../AoC2020/solutions/09";
size_t first = 0;

void task_01(){
    auto data = loadTokens<size_t>(sourceDirectory + "/input");
    unsigned range = 25;
    for (unsigned i = 0; i < data.size(); ++i){
        bool found = false;
        for (unsigned x = 0; x < range && !found; ++x){
            for (unsigned y = 0; y < range && !found; ++y) {
                auto dx=data[i + x], dy=data[i + y];
                if (dx != dy){
                    if (dx+dy == data[i+range]){
                        found = true;
                    }
                }
            }
        }
        if(!found){
            first = data[i + range];
            print_solution(9, true, first);
            return;
        }
    }

}

void task_02(){
    auto data = loadTokens<size_t>(sourceDirectory + "/input");
    for (auto i = data.begin(); i != data.end(); ++i){
        size_t sum = 0;
        for (unsigned j = 0;; ++j){
            sum += i[j];
            if (sum > first){
                break;
            }
            if (sum == first){
                auto min = *std::min_element(i, i+j+1);
                auto max = *std::max_element(i, i+j+1);
                print_solution(9, false, min+max);
                return;
            }
        }
    }

}

int main(){
    task_01();
    task_02();
}

#endif //AOC_09_H
